pkg/controller/daemonset/update.go

```diff
@@ -0,0 +1,708 @@
+			// respect partitioned nodes to keep old versions.
+			nodeToDaemonPods = dsc.getNodeToDaemonPodsByPartition(ds, nodeToDaemonPods)
+		}
+	} else {
+		// respect selected nodes to update.
+		nodeToDaemonPods = dsc.getNodeToDaemonPodsBySelector(ds, nodeToDaemonPods)
+	}
+
+	maxSurge, numSurge, err := dsc.getSurgeNumbers(ds, nodeToDaemonPods, hash)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("Couldn't get surge numbers: %v", err)
+	}
+
+	nodesWantToRun, nodesShouldContinueRunning, err := dsc.getNodesShouldRunDaemonPod(ds)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("Couldn't get nodes which want to run ds pod: %v", err)
+	}
+
+	var nodesToSurge []string
+	var oldPodsToDelete []string
+	for node, pods := range nodeToDaemonPods {
+		var newPod, oldPod *corev1.Pod
+		wantToRun := nodesWantToRun.Has(node)
+		shouldContinueRunning := nodesShouldContinueRunning.Has(node)
+		// if node has new taint, then the already existed pod does not want to run but should continue running
+		if !wantToRun && !shouldContinueRunning {
+			for _, pod := range pods {
+				klog.V(6).Infof("Marking pod %s/%s on unsuitable node %s for deletion", ds.Name, pod.Name, node)
+				if pod.DeletionTimestamp == nil {
+					oldPodsToDelete = append(oldPodsToDelete, pod.Name)
+				}
+			}
+			continue
+		}
+		foundAvailable := false
+		for _, pod := range pods {
+			generation, err := GetTemplateGeneration(ds)
+			if err != nil {
+				generation = nil
+			}
+			if util.IsPodUpdated(pod, hash, generation) {
+				if newPod != nil {
+					klog.Warningf("Multiple new pods on node %s: %s, %s", node, newPod.Name, pod.Name)
+				}
+				newPod = pod
+			} else {
+				if oldPod != nil {
+					klog.Warningf("Multiple old pods on node %s: %s, %s", node, oldPod.Name, pod.Name)
+				}
+				oldPod = pod
+			}
+			if !foundAvailable && podutil.IsPodAvailable(pod, ds.Spec.MinReadySeconds, metav1.Now()) {
+				foundAvailable = true
+			}
+		}
+
+		if newPod != nil {
+			klog.Infof("newPod IsPodAvailable is %v", podutil.IsPodAvailable(newPod, ds.Spec.MinReadySeconds, metav1.Now()))
+		}
+		if newPod == nil && numSurge < maxSurge && wantToRun {
+			if !foundAvailable && len(pods) >= 2 {
+				klog.Warningf("Node %s already has %d unavailable pods, need clean first, skip surge new pod", node, len(pods))
+			} else {
+				klog.V(6).Infof("Surging new pod on node %s", node)
+				numSurge++
+				nodesToSurge = append(nodesToSurge, node)
+			}
+		} else if newPod != nil && podutil.IsPodAvailable(newPod, ds.Spec.MinReadySeconds, metav1.Now()) && oldPod != nil && oldPod.DeletionTimestamp == nil {
+			klog.V(6).Infof("Marking pod %s/%s for deletion", ds.Name, oldPod.Name)
+			oldPodsToDelete = append(oldPodsToDelete, oldPod.Name)
+		} else {
+			if ds.Spec.MinReadySeconds > 0 {
+				// TODO: use return reconcile.Result{RequeueAfter: time.Duration(ds.Spec.MinReadySeconds) * time.Second}, nil
```

Comment:
I think we should not sleep here.

Reply:
yes, the sleep is so ugly, and i remove it at once.